presto-main/src/test/java/com/facebook/presto/execution/TestThriftTaskStatus.java

```diff
@@ -0,0 +1,209 @@
+
+        ExecutionFailureInfo secondFailure = failures.get(1);
+        assertEquals(secondFailure.getType(), PrestoTransportException.class.getName());
+        assertEquals(secondFailure.getMessage(), "Too many requests failed");
+        assertEquals(secondFailure.getRemoteHost(), REMOTE_HOST);
+        assertEquals(secondFailure.getErrorCode(), TOO_MANY_REQUESTS_FAILED.toErrorCode());
+        ExecutionFailureInfo cause = secondFailure.getCause();
+        assertEquals(cause.getType(), PrestoException.class.getName());
+        assertEquals(cause.getMessage(), "Remote Task Error");
+        assertEquals(cause.getErrorCode(), REMOTE_TASK_ERROR.toErrorCode());
+
+        ExecutionFailureInfo thirdFailure = failures.get(2);
+        assertEquals(thirdFailure.getType(), ParsingException.class.getName());
+        assertEquals(thirdFailure.getErrorCode(), SYNTAX_ERROR.toErrorCode());
+        assertEquals(thirdFailure.getErrorLocation().getLineNumber(), 100);
+        assertEquals(thirdFailure.getErrorLocation().getColumnNumber(), 3);
+    }
+
+    private TaskStatus getRoundTripSerialize(ThriftCodec<TaskStatus> readCodec, ThriftCodec<TaskStatus> writeCodec, Function<TTransport, TProtocol> protocolFactory)
+            throws Exception
+    {
+        TProtocol protocol = protocolFactory.apply(transport);
+        writeCodec.write(taskStatus, protocol);
+        return readCodec.read(protocol);
+    }
+
+    private TaskStatus getTaskStatus()
+    {
+        List<ExecutionFailureInfo> executionFailureInfos = getExecutionFailureInfos();
+        return new TaskStatus(
+                TASK_INSTANCE_ID_LEAST_SIGNIFICANT_BITS,
+                TASK_INSTANCE_ID_MOST_SIGNIFICANT_BITS,
+                VERSION,
+                RUNNING,
+                SELF_URI,
+                LIFESPANS,
+                executionFailureInfos,
+                QUEUED_PARTITIONED_DRIVERS,
+                RUNNING_PARTITIONED_DRIVERS,
+                OUTPUT_BUFFER_UTILIZATION,
+                OUTPUT_BUFFER_OVERUTILIZED,
+                PHYSICAL_WRITTEN_DATA_SIZE_IN_BYTES,
+                MEMORY_RESERVATION_IN_BYTES,
+                SYSTEM_MEMORY_RESERVATION_IN_BYTES,
+                FULL_GC_COUNT,
+                FULL_GC_TIME_IN_MILLIS);
+    }
+
+    private List<ExecutionFailureInfo> getExecutionFailureInfos()
+    {
+        IOException ioException = new IOException("Remote call timed out");
+        ioException.addSuppressed(new IOException("Thrift call timed out"));
+        PrestoTransportException prestoTransportException = new PrestoTransportException(TOO_MANY_REQUESTS_FAILED,
+                REMOTE_HOST,
+                "Too many requests failed",
+                new PrestoException(REMOTE_TASK_ERROR, "Remote Task Error"));
+        ParsingException parsingException = new ParsingException("Parsing Exception", new NodeLocation(100, 1));
+        return Failures.toFailures(ImmutableList.of(ioException, prestoTransportException, parsingException));
```

Comment:
Are we sure about covering the case when ExecutionFailureInfo is recursive?

Reply:
The `PrestoTransportException` takes in a `PrestoException` as a throwable which becomes the cause and starts setting up the recursive structure which we assert in these lines :

```
        ExecutionFailureInfo secondFailure = failures.get(1);
        assertEquals(secondFailure.getType(), PrestoTransportException.class.getName());
        assertEquals(secondFailure.getMessage(), "Too many requests failed");
        assertEquals(secondFailure.getRemoteHost(), REMOTE_HOST);
        assertEquals(secondFailure.getErrorCode(), TOO_MANY_REQUESTS_FAILED.toErrorCode());
        //The recursive structure
        ExecutionFailureInfo cause = secondFailure.getCause();
        assertEquals(cause.getType(), PrestoException.class.getName());
        assertEquals(cause.getMessage(), "Remote Task Error");
        assertEquals(cause.getErrorCode(), REMOTE_TASK_ERROR.toErrorCode());
```